subcerts.go

```diff
@@ -0,0 +1,508 @@
+	case *ecdsa.PublicKey:
+		curveName := pk.Curve.Params().Name
+		if curveName == "P-256" {
+			return pk, ECDSAWithP256AndSHA256, nil
+		} else if curveName == "P-384" {
+			return pk, ECDSAWithP384AndSHA384, nil
+		} else if curveName == "P-521" {
+			return pk, ECDSAWithP521AndSHA512, nil
+		} else {
+			return nil, 0, fmt.Errorf("curve %s s not supported", curveName)
+		}
+
+	default:
+		return nil, 0, fmt.Errorf("unsupported delgation key type: %T", pk)
+	}
+}
+
+// marshal encodes a credential as per the spec.
+func (cred *credential) marshal() ([]byte, error) {
+	// Write the valid_time field.
+	serialized := make([]byte, 6)
+	binary.BigEndian.PutUint32(serialized, uint32(cred.validTime/time.Second))
+
+	// Encode the public key and assert that the encoding is no longer than 2^16
+	// bytes (per the spect).
+	serializedPublicKey, err := cred.marshalSubjectPublicKeyInfo()
+	if err != nil {
+		return nil, err
+	}
+	if len(serializedPublicKey) > dcMaxPublicKeyLen {
+		return nil, errors.New("public key is too long")
+	}
+
+	// Write the length of the public_key field.
+	binary.BigEndian.PutUint16(serialized[4:], uint16(len(serializedPublicKey)))
+
+	// Write the public key.
+	return append(serialized, serializedPublicKey...), nil
+}
+
+// unmarshalCredential decodes a credential and returns it.
+func unmarshalCredential(serialized []byte) (*credential, error) {
+	// Check that the serialized credential is long enough and compute the
+	// length of the serialized public key.
+	serializedPublicKeyLen, err := getCredentialLen(serialized)
+	if err != nil {
+		return nil, err
+	}
+	serializedPublicKeyLen -= 6
+
+	// Parse the validity time.
+	validTime := time.Duration(binary.BigEndian.Uint32(serialized)) * time.Second
+
+	// Parse the SubjectPublicKeyInfo.
+	serialized = serialized[6:]
+	pk, scheme, err := unmarshalSubjectPublicKeyInfo(serialized)
+	if err != nil {
+		return nil, err
+	}
+
+	return &credential{validTime, pk, scheme}, nil
+}
+
+// getCredentialLen returns the number of bytes comprising the serialized
+// credential that starts at the beginning of the input slice. It returns an
+// error if the input is too short to contain a credential.
+func getCredentialLen(serialized []byte) (int, error) {
+	if len(serialized) < 6 {
+		return 0, errors.New("credential is too short")
+	}
+	// First 4 bytes is the validity time.
+	serialized = serialized[4:]
+
+	// The next 2 bytes are the length of the serialized public key.
+	serializedPublicKeyLen := int(binary.BigEndian.Uint16(serialized))
+	serialized = serialized[2:]
+
+	if len(serialized) < serializedPublicKeyLen {
+		return 0, errors.New("public key of credential is too short")
+	}
+
+	return 6 + serializedPublicKeyLen, nil
+}
+
+// DelegatedCredential stores a credential and its delegation.
+type DelegatedCredential struct {
+	// The serialized form of the credential.
+	Raw []byte
+
+	// Time in nonsecnds for which the credential is valid. When this data
```

Comment:
~~nanoseconds~~ Specifying unit shouldn't be necessary, b/c its a time.Duration. "ValidTime is how long the cred is valid"

Reply:
Removed "in nanoseconds".